File: js/1.6.js
const EPSILON = 0.001

const newIf = (cond, then, els) => cond() ? then : els

const square = x => x * x

const avg = (x, y) => (x + y) / 2

const isGoodEnough = (guess, x) => Math.abs(x - square(guess)) < EPSILON

const improve = (guess, x) => avg(x / guess, guess)

const sqrtIter = (guess, x) => {
    newIf(isGoodEnough(guess, x), guess, sqrtIter(improve(guess, x), x))
}

const sqrt = x => sqrtIter(1, x)

console.log(sqrt(2))

// same thing as in 1.6.clj. Infinite loop because `newIf` is not a special form,
// hence according to applicative-order eval all arguments are evaluated.
// `if` is a special form and do not evaluate the code which was not reached

File: 1.6.clj
; Why `if` is a special form?  try custom new-if and see
(defn new-if [predicate then-clause else-clause] (cond predicate then-clause
                                                       :else else-clause))
(defn abs [x] (if (< x 0) (* -1 x) x))

(defn square [x] (* x x))

(defn good-enough? [guess x] (< (abs (- (square guess) x)) 0.001))

(defn average [x y] (/ (+ x y) 2.0))

(defn improve [guess x] (average guess (/ x guess)))

(defn sqrt-iter [guess x]
  (new-if (good-enough? guess x)
          guess
          (sqrt-iter (improve guess x) x)))

(defn sqrt [x] (sqrt-iter 1 x))

(println (sqrt 2))

; it stucks in infinite loop because special form `if` execute one path or another based on a predicate
; `new-if` uses standard applicative-order evaluation and evaluates all arguments no matter what predicate is